rasa/core/policies/memoization.py

```diff
@@ -279,25 +280,43 @@ def _back_to_the_future_again(tracker) -> Optional[DialogueStateTracker]:
                     idx_of_second_action = e_i
                     break
 
-        if idx_of_second_action is None:
-            return None
+        # use first action, if we went first time and second action, if we went again
+        idx_to_use = idx_of_second_action if again else idx_of_first_action
+        if idx_to_use is None:
+            return
+
         # make second ActionExecuted the first one
-        events = tracker.applied_events()[idx_of_second_action:]
+        events = applied_events[idx_to_use:]
         if not events:
-            return None
+            return
 
         mcfly_tracker = tracker.init_copy()
         for e in events:
             mcfly_tracker.update(e)
 
         return mcfly_tracker
 
-    def _recall_using_delorean(self, old_states, tracker, domain) -> Optional[int]:
-        """Recursively go to the past to correctly forget slots,
-            and then back to the future to recall."""
+    def _recall_using_delorean(
+        self, old_states: List, tracker: DialogueStateTracker, domain: Domain,
+    ) -> Optional[int]:
+        """Applies to the future idea to change the past and get the new future.
```

Comment:
what does this sentence mean? 🙈

Reply:
It is directly copied from main. I'd rewrite much of the names and comments, but this is an old version